roles/rsyslog/tasks/inputs/ovirt/set_vars.yml

```diff
@@ -0,0 +1,21 @@
+---
+- name: Ensure ansible_facts used by role
+  setup:
+    gather_subset:
+      - 'min'
+      - default_ipv4
```

Comment:
You don't need to quote `min` here.  Also, `default_ipv4` should go under `filter` - it isn't a `gather_subset` - https://docs.ansible.com/ansible/latest/collections/ansible/builtin/setup_module.html#setup-module
also, if `default_ipv4` is included with `min`, then you don't need to specify it

Reply:
Is this your suggestion?
```
diff --git a/roles/rsyslog/tasks/inputs/ovirt/set_vars.yml b/roles/rsyslog/tasks/inputs/ovirt/set_vars.yml
index d6bdcfe..686aeb6 100644
--- a/roles/rsyslog/tasks/inputs/ovirt/set_vars.yml
+++ b/roles/rsyslog/tasks/inputs/ovirt/set_vars.yml
@@ -2,7 +2,8 @@
 - name: Ensure ansible_facts used by role
   setup:
     gather_subset:
-      - 'min'
+      - min
+    filter:
       - default_ipv4
   when: not ansible_facts.keys() | list |
     intersect(__logging_required_facts) == __logging_required_facts
```
If so, it makes the ovirt input fail due to missing ansible_default_ipv4...
```
  set $!ipaddr4 = '{{ ansible_default_ipv4.address }}';
```
Please note that `min` does not include `default_ipv4`.

So far, only this param set works for me.
```
  setup:
    gather_subset:
      - min
      - default_ipv4
```

Reply:
ok - using `default_ipv4` in gather_subset is definitely not documented - looks like you should use `network` to get ipv4 and other network facts.

Reply:
so try
```yaml
 - name: Ensure ansible_facts used by role
   setup:
     gather_subset:
       - min
       - network
   when: not ansible_facts.keys() | list |
     intersect(__logging_required_facts) == __logging_required_facts

Reply:
Hmmm, I thought we want to gather facts as less as possible. So, I put `default_ipv4` instead of `network`. I guess my assumption was wrong?

Reply:
> Hmmm, I thought we want to gather facts as less as possible. So, I put `default_ipv4` instead of `network`. I guess my assumption was wrong?

The way it was explained to me by one of the Ansible guys is that gather_subset is the name of a collection of facts - for example, `min` uses these subsets - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/setup.py#L192-L195 - `network` is another subset.  Specifying the fewest subsets possible will speed up Ansible processing.  You can also use `filter` to reduce the amount of network traffic by specifying only those parameters you want to be returned.  In the case of logging, you could use `filter` and specify the list of `default_ipv4`, `distribution`, etc. so that only those values will be returned.  AFAICT by looking at the ansible code, `default_ipv4` is *not* a fact subset, it is a parameter, so it should not be used in `gather_subset`, only in `filter`, and I'm not sure why it works, but since this is undocumented behavior, we should not use it.  So to be maximally performant, we should use something like this:
```yaml
 - name: Ensure ansible_facts used by role
   setup:
     gather_subset:
       - min
       - network
     filter: "{{ ['default_ipv4'] + __logging_required_facts }}"
   when: not ansible_facts.keys() | list |
     intersect(__logging_required_facts) == __logging_required_facts
```
that is - gather the minimal and network facts, and only return the values specified in the `filter` list, which are the ones needed by the role.

I'm not sure how much of a performance gain it is to specify `filter`, but in this case, it is pretty straightforward to specify the list of parameters.

One problem is that I don't know if specifying a list of parameters to `filter` is compatible with 2.9.  The documentation says this: https://docs.ansible.com/ansible/2.9/modules/setup_module.html#setup-module
which implies that `filter` is a string, not a list.  I didn't look at the 2.9 code, but the 2.12 code will accept `filter` as a list or as a string.  This is one of the reasons I didn't want to use `filter` in general, because the API is unclear.  However, if you want to do some investigation to determine if it is safe to use a list for 2.9, then we can use `filter` with a list.

Reply:
Thank you for your detailed explanation, @richm. You are right. `gather_subset` is supposed take only specific values... Somehow I thought it also took individual fact (and it's unfortunate it "accidentally" does...) Since it is not documented, it could stop working any time. So, I should replace it with `network`. And I don't think the `filter` part is needed. The reason I thought having `default_ipv4` was a good idea was it could just retrieve the ipv4 fact, instead of all the `network` facts. Retrieving all then dropping the rest does not provide us much benefit, I guess.